cmd/integration_test.go

```diff
@@ -669,46 +666,80 @@ func TestAbortedByUserWithGoodThresholds(t *testing.T) {
 	assert.Contains(t, stdOut, `level=debug msg="Sending test finished" output=cloud ref=111 run_status=5 tainted=false`)
 }
 
-func TestAbortedByUserWithRestAPI(t *testing.T) {
-	t.Parallel()
-	script := []byte(`
-		import { sleep } from 'k6';
-		export default function () {
-			console.log('a simple iteration')
-			sleep(1);
-		};
-
-		export function teardown() {
-			console.log('teardown() called');
-		}
-	`)
-
-	ts := getSimpleCloudOutputTestState(
-		t, script, []string{"-v", "--log-output=stdout", "--iterations", "20"},
-		lib.RunStatusAbortedUser, cloudapi.ResultStatusPassed, 0,
-	)
-
-	wg := sync.WaitGroup{}
+func asyncWaitForStdoutAndRun(
+	t *testing.T, ts *globalTestState, attempts int, interval time.Duration, expText string, callback func(),
+) {
+	wg := &sync.WaitGroup{}
 	wg.Add(1)
-
 	go func() {
 		defer wg.Done()
-		newRootCommand(ts.globalState).execute()
-	}()
+		reachedCondition := false
+		for i := 0; i < attempts; i++ {
+			ts.outMutex.Lock()
+			stdOut := ts.stdOut.String()
+			ts.outMutex.Unlock()
+
+			if strings.Contains(stdOut, expText) {
+				t.Logf("found '%s' in the process stdout on try %d at t=%s", expText, i, time.Now())
+				reachedCondition = true
+				break
+			}
+
+			t.Logf("did not find the text '%s' in the process stdout on try %d at t=%s", expText, i, time.Now())
+			time.Sleep(interval)
+		}
+		if reachedCondition {
+			callback()
+			return // everything is fine
+		}
 
-	reachedIteration := false
-	for i := 0; i <= 10 && reachedIteration == false; i++ {
-		time.Sleep(1 * time.Second)
 		ts.outMutex.Lock()
 		stdOut := ts.stdOut.String()
 		ts.outMutex.Unlock()
+		t.Log(stdOut)
+		require.FailNow(
+			t, "did not find the text '%s' in the process stdout after %d attempts (%s)",
+			expText, attempts, time.Duration(attempts)*interval,
+		)
+	}()
 
-		if !strings.Contains(stdOut, "a simple iteration") {
-			t.Logf("did not see an iteration on try %d at t=%s", i, time.Now())
-			continue
+	t.Cleanup(wg.Wait) // ensure the test waits for the goroutine to finish
+}
+
+func asyncWaitForStdoutAndStopTestWithInterruptSignal(
+	t *testing.T, ts *globalTestState, attempts int, interval time.Duration, expText string,
+) {
+	sendSignal := make(chan struct{})
+	ts.globalState.signalNotify = func(c chan<- os.Signal, signals ...os.Signal) {
+		isAbortNotify := false
+		for _, s := range signals {
+			if s == os.Interrupt {
+				isAbortNotify = true
+				break
+			}
+		}
+		if !isAbortNotify {
+			return
 		}
+		go func() {
+			<-sendSignal
+			c <- os.Interrupt
+			close(sendSignal)
+		}()
+	}
+	ts.globalState.signalStop = func(c chan<- os.Signal) { /* noop */ }
 
-		reachedIteration = true
+	asyncWaitForStdoutAndRun(t, ts, attempts, interval, expText, func() {
+		t.Log("expected stdout text was found, sending interrupt signal...")
+		sendSignal <- struct{}{}
+		<-sendSignal
+	})
+}
+
+func asyncWaitForStdoutAndStopTestFromRESTAPI(
```

Comment:
This function is only called from `TestAbortedByUserWithRestAPI`, and just wraps another helper function. So I would inline the `asyncWaitForStdoutAndRun()` call, and remove this.

`asyncWaitForStdoutAndRun()` is also called with the same arguments, except for the callback func. So I would drop the unneded arguments, and add them once they're needed.

In general, abstractions add a layer of indirection that complicates code comprehension, so I wouldn't use them so liberally, and would prefer even some duplication between tests. Or, if tests really are similar, turn them into table tests and avoid the helper functions.

Reply:
I agree and I was about to fix this, but I think the fix here is to just add more tests that abort the test run via the k6 REST API :sweat_smile: See https://github.com/grafana/k6/issues/2804, we need more tests about this behavior... :disappointed: 

In https://github.com/grafana/k6/pull/2815 or soon after it, we should be able to nicely abort the test with the REST even during VU init or `setup()`, so I'd prefer to keep this helper method here and add more tests in that PR.